c:geo not starting after restore of data

While testing other issues I ran into a problem after this scenario:

- 2021.03.01-NB
- Go to device settings - Apps - c:geo and delete all data for c:geo
- Start c:geo (comes up with wizard)
- Complete the wizard (permissions, base folder, login to geocaching.com, skip advanced wizard)
- c:geo works normally
- Go to c:geo menu - Settings - Restore and select to restore caches and settings (used a backup done with yesterdays nightly, settings did not include account data)
- Backup is restored and restart is triggered by dialog

Result:
- Grey screen is shown, after some time splash screen logo shows up
- App runs into ANR
- Consecutive tries to start c:geo end up in ANR

Not yer knowing the exact cause of this problem, at least the following came to my mind:
What happens to previously existing SAF persistent folders in such case? The settings are restored, but the permissions might be lost...or is this also covered by the backup?

I ran into a similar problem in the emulator. There is a stack trace, but no other SAF related logging in the logcat...

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 19449
    java.lang.ExceptionInInitializerError
        at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61)
        at cgeo.geocaching.utils.Log.<clinit>(Log.java:51)
        at cgeo.geocaching.utils.Log.setDebug(Log.java:87)
        at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118)
        at cgeo.geocaching.settings.Settings.getApplicationLocale(Settings.java:1535)
        at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116)
        at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1182)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460)
        at android.app.ActivityThread.access$1300(ActivityThread.java:219)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri cgeo.geocaching.storage.ContentStorage.getUriForFolder(cgeo.geocaching.storage.Folder)' on a null object reference
        at cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:73)
        at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:548)
        at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:542)
        at cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:505)
        at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:116)
        at cgeo.geocaching.storage.ContentStorage.reevaluateFolderDefaults(ContentStorage.java:110)
        at cgeo.geocaching.storage.ContentStorage.<init>(ContentStorage.java:102)
        at cgeo.geocaching.storage.ContentStorage.<clinit>(ContentStorage.java:58)
        at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61) 
        at cgeo.geocaching.utils.Log.<clinit>(Log.java:51) 
        at cgeo.geocaching.utils.Log.setDebug(Log.java:87) 
        at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118) 
        at cgeo.geocaching.settings.Settings.getApplicationLocale(Settings.java:1535) 
        at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116) 
        at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52) 
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1182) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460) 
        at android.app.ActivityThread.access$1300(ActivityThread.java:219) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
```

after several hours I checked again anf it meanwhile started.

Log is full of this error:
``
03-02 14:09:57.035 27292 27292 E AndroidRuntime: FATAL EXCEPTION: main
03-02 14:09:57.035 27292 27292 E AndroidRuntime: Process: cgeo.geocaching, PID: 27292
03-02 14:09:57.035 27292 27292 E AndroidRuntime: java.lang.ExceptionInInitializerError
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.<clinit>(Log.java:51)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.setDebug(Log.java:87)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.handleMakeApplication(ActivityThread.java:7501)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7440)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.access$1500(ActivityThread.java:301)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2148)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:246)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8506)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri cgeo.geocaching.storage.ContentStorage.getUriForFolder(cgeo.geocaching.storage.Folder)' on a null object reference
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:73)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:548)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:542)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:505)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:116)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reevaluateFolderDefaults(ContentStorage.java:110)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<init>(ContentStorage.java:102)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<clinit>(ContentStorage.java:58)
03-02 14:09:57.035 27292 27292 E AndroidRuntime: 	... 17 more
03-02 14:10:28.489 27787 27787 E AndroidRuntime: FATAL EXCEPTION: main
03-02 14:10:28.489 27787 27787 E AndroidRuntime: Process: cgeo.geocaching, PID: 27787
03-02 14:10:28.489 27787 27787 E AndroidRuntime: java.lang.ExceptionInInitializerError
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.get(ContentStorage.java:61)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.<clinit>(Log.java:51)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.utils.Log.setDebug(Log.java:87)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.settings.Settings.<clinit>(Settings.java:118)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.initApplicationLocale(CgeoApplication.java:116)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.CgeoApplication.onCreate(CgeoApplication.java:52)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.handleMakeApplication(ActivityThread.java:7501)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7440)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.access$1500(ActivityThread.java:301)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2148)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:246)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8506)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.net.Uri cgeo.geocaching.storage.ContentStorage.getUriForFolder(cgeo.geocaching.storage.Folder)' on a null object reference
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.utils.LocalizationUtils.getMultiPurposeString(LocalizationUtils.java:73)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:548)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reportProblem(ContentStorage.java:542)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.getFolder(ContentStorage.java:505)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.ensureFolder(ContentStorage.java:116)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.reevaluateFolderDefaults(ContentStorage.java:110)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<init>(ContentStorage.java:102)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	at cgeo.geocaching.storage.ContentStorage.<clinit>(ContentStorage.java:58)
03-02 14:10:28.489 27787 27787 E AndroidRuntime: 	... 17 more
``

--- a/main/src/cgeo/geocaching/maps/mapsforge/v6/RenderThemeHelper.java
+++ b/main/src/cgeo/geocaching/maps/mapsforge/v6/RenderThemeHelper.java
@@ -464,7 +464,7 @@ public class RenderThemeHelper implements XmlRenderThemeMenuCallback, SharedPref
         public static MapThemeFolderSyncTask createAndExecute(final Folder source, final File target, final Consumer<FolderUtils.FolderProcessResult> callback) {
             final MapThemeFolderSyncTask task = new MapThemeFolderSyncTask(source, target, callback);
             task.execute();
-
+            Log.i("[MapThemeFolderSync] start synchronization " + source + " -> " + target);
             return task;
         }
 
@@ -480,6 +480,7 @@ public class RenderThemeHelper implements XmlRenderThemeMenuCallback, SharedPref
                 if (taskIsDone) {
                     return false;
                 }
+                Log.i("[MapThemeFolderSync] REstart synchronization " + source + " -> " + target);
                 cancelFlag.set(true);
                 requestRedo = true;
                 startTime = System.currentTimeMillis();
@@ -505,6 +506,7 @@ public class RenderThemeHelper implements XmlRenderThemeMenuCallback, SharedPref
 
         @Override
         protected void onPostExecute(final FolderUtils.FolderProcessResult result) {
+            Log.i("[MapThemeFolderSync] Finished synchronization ");
             //show toast only if something actually happened
             if (result.filesModified > 0) {
                 showToast(R.string.mapthemes_foldersync_finished_toast,
@@ -515,6 +517,7 @@ public class RenderThemeHelper implements XmlRenderThemeMenuCallback, SharedPref
             if (callback != null) {
                 callback.accept(result);
             }
+            Log.i("[MapThemeFolderSync] Finished synchronization callback");
         }
 
         private static void showToast(final int resId, final Object ... params) {

--- a/main/src/cgeo/geocaching/storage/ContentStorage.java
+++ b/main/src/cgeo/geocaching/storage/ContentStorage.java
@@ -106,14 +106,15 @@ public class ContentStorage {
         for (PersistableFolder folder : PersistableFolder.values()) {
             //(re)sets default folders and ensures that it is definitely accessible
             folder.setDefaultFolder(getAccessibleDefaultFolder(folder.getDefaultCandidates(), folder.needsWrite(), folder.name()));
-            //tests user-defined folder (if any) and reset to default if not accessible
-            ensureFolder(folder);
         }
     }
 
     /** checks if folder is available and can be used, creates it if need be. */
     public boolean ensureFolder(final PersistableFolder publicFolder) {
         final Folder folder = getFolder(publicFolder);
+        if (folder == null) {
+            return false;
+        }
         return ensureFolder(folder, publicFolder.needsWrite(), false);
      }
 
@@ -257,6 +258,10 @@ public class ContentStorage {
         } else {
             parentFolder = Folder.fromFolder(rootFolder, subfolderAndName.substring(0, idx));
             fileName = subfolderAndName.substring(idx + 1);
+        }
+
+        if (parentFolder == null) {
+            return null;
         }
 
         try {

--- a/main/src/cgeo/geocaching/utils/LocalizationUtils.java
+++ b/main/src/cgeo/geocaching/utils/LocalizationUtils.java
@@ -64,13 +64,14 @@ public final class LocalizationUtils {
         //prepare params message
         final Object[] paramsForLog = new Object[params.length];
         final Object[] paramsForUser = new Object[params.length];
+        //Note that ContentStorage.get() can actually be null here in case there was an error in initialization of Log!
         for (int i = 0; i < params.length; i++) {
             if (params[i] instanceof Folder) {
                 paramsForUser[i] = ((Folder) params[i]).toUserDisplayableString();
-                paramsForLog[i] = params[i] + "(" + ContentStorage.get().getUriForFolder((Folder) params[i]) + ")";
+                paramsForLog[i] = params[i] + "(" + ContentStorage.get() == null ? null : ContentStorage.get().getUriForFolder((Folder) params[i]) + ")";
             } else if (params[i] instanceof PersistableFolder) {
                 paramsForUser[i] = ((PersistableFolder) params[i]).toUserDisplayableValue();
-                paramsForLog[i] = params[i] + "(" + ContentStorage.get().getUriForFolder(((PersistableFolder) params[i]).getFolder()) + ")";
+                paramsForLog[i] = params[i] + "(" + ContentStorage.get() == null ? null : ContentStorage.get().getUriForFolder(((PersistableFolder) params[i]).getFolder()) + ")";
             } else if (params[i] instanceof Uri) {
                 paramsForUser[i] = UriUtils.toUserDisplayableString((Uri) params[i]);
                 paramsForLog[i] = params[i];

--- a/main/src/cgeo/geocaching/utils/Log.java
+++ b/main/src/cgeo/geocaching/utils/Log.java
@@ -61,10 +61,11 @@ public final class Log {
                 setProperties(logProps);
             }
         } catch (Exception ex) {
+            //whatever happens in Log initializer, it is NOT allowed to make Log unusable!
             android.util.Log.e(TAG, "[Log] Failed to set up Logging", ex);
         } finally {
-                IOUtils.closeQuietly(propFile);
-            }
+            IOUtils.closeQuietly(propFile);
+        }
     }
 
     private Log() {
@@ -102,7 +103,7 @@ public final class Log {
             }
             if (logProps.containsKey(PROP_CALLERINFO_MAXDEPTH)) {
                 try {
-                    addCallerInfoMaxDepth = Integer.valueOf(logProps.getProperty(PROP_CALLERINFO_MAXDEPTH));
+                    addCallerInfoMaxDepth = Integer.parseInt(logProps.getProperty(PROP_CALLERINFO_MAXDEPTH));
                 } catch (NumberFormatException nfe) {
                     //no valid maxDepth in prop file, ignore this
                 }

